src/constraint.rs

```diff
@@ -777,3 +791,69 @@ impl ConstraintValidator for ContainsConstraint {
         Ok(())
     }
 }
+
+/// Implements an `container_length` constraint of Ion Schema
+/// [container_length]: https://amzn.github.io/ion-schema/docs/spec.html#container_length
+#[derive(Debug, Clone, PartialEq)]
+pub struct ContainerLengthConstraint {
+    isl_length: IslLength,
+}
+
+impl ContainerLengthConstraint {
+    pub fn new(isl_length: IslLength) -> Self {
+        Self { isl_length }
+    }
+
+    pub fn isl_length(&self) -> &IslLength {
+        &self.isl_length
+    }
+}
+
+impl ConstraintValidator for ContainerLengthConstraint {
+    fn validate(&self, value: &OwnedElement, type_store: &TypeStore) -> ValidationResult {
+        // Check for null container
+        if value.is_null() {
+            return Err(Violation::new(
+                "container_length",
+                ViolationCode::TypeMismatched,
+                &format!("expected a container found {:?}", value),
+            ));
+        }
+
+        // get the size of given value container
+        let size = match value.ion_type() {
+            IonType::List => value.as_sequence().unwrap().iter().count(),
+            IonType::SExpression => value.as_sequence().unwrap().iter().count(),
+            IonType::Struct => value.as_struct().unwrap().iter().count(),
+            _ => {
+                // return Violation if value is not an Ion container
+                return Err(Violation::new(
+                    "container_length",
+                    ViolationCode::TypeMismatched,
+                    &format!(
+                        "expected list/sexp/struct(container type) found {}",
+                        value.ion_type()
+                    ),
+                ));
+            }
+        };
+
+        // get isl length as a range
+        // TODO: include IonSchemaError in Violations to remove usage of expect
+        let isl_length: Range = self
+            .isl_length()
+            .try_into()
+            .expect("Unable to parse container_length to range");
```

Comment:
I did realize that this will be incorrect to pass on `IonSchemaError` in validation. I have 2 options here to implement it correctly:
1. Change `TryFrom<IslOccurs>/ `TryFrom<IslLength>` (which return `IonSchemaResult<Range>`)  to be `From<>` (which return `Range` itself) as at this point the range was already parsed properly.
2. Remove usage of `IslOccurs` / `IslLength` and just directly pass in a `Range` to the constraints. I created these implementations because I am trying to keep `IslOccurs` in the ISL phase because it aage of `IslOccurs` / `IslLength` and just directly pass in a `Range` to the constraints. In my understandinligns with Ion Schema specification grammar and it simplifies for users to create occurs constraint programmatically using ISL APIs. Being said that yes it possible to remove both IslOccurs and IslLength and just pass in the Range to constraints.

Looking at @zslayton and your comment I think there is more inclination towards 2nd option. I will create an issue for it and do it in separate PR if that sounds good to both of you.